RecoHGCal/TICL/plugins/TracksterLinksProducer.cc

```diff
@@ -0,0 +1,258 @@
+  }
+  LogDebug("TracksterLinksProducer") << std::endl;
+}
+
+void TracksterLinksProducer::produce(edm::Event &evt, const edm::EventSetup &es) {
+  auto resultTracksters = std::make_unique<std::vector<Trackster>>();
+
+  auto linkedResultTracksters = std::make_unique<std::vector<std::vector<unsigned int>>>();
+
+  const auto &layerClusters = evt.get(clusters_token_);
+  const auto &layerClustersTimes = evt.get(clustersTime_token_);
+
+  // loop over the original_masks_tokens_ and get the original masks collections and multiply them
+  // to get the global mask
+  std::vector<float> original_global_mask(layerClusters.size(), 1.f);
+  for (unsigned int i = 0; i < original_masks_tokens_.size(); ++i) {
+    const auto &tmp_mask = evt.get(original_masks_tokens_[i]);
+    for (unsigned int j = 0; j < tmp_mask.size(); ++j) {
+      original_global_mask[j] *= tmp_mask[j];
+    }
+  }
+
+  auto resultMask = std::make_unique<std::vector<float>>(original_global_mask);
+
+  std::vector<edm::Handle<std::vector<Trackster>>> tracksters_h(tracksters_tokens_.size());
+  MultiVectorManager<Trackster> trackstersManager;
+  for (unsigned int i = 0; i < tracksters_tokens_.size(); ++i) {
+    evt.getByToken(tracksters_tokens_[i], tracksters_h[i]);
+    //Fill MultiVectorManager
+    trackstersManager.addVector(*tracksters_h[i]);
+  }
+
+  // Linking
+  const typename TracksterLinkingAlgoBase::Inputs input(evt, es, layerClusters, layerClustersTimes, trackstersManager);
+  std::vector<std::vector<unsigned int>> linkedTracksterIdToInputTracksterId;
+
+  // LinkTracksters will produce a vector of vector of indices of tracksters that:
+  // 1) are linked together if more than one
+  // 2) are isolated if only one
+  // Result tracksters contains the final version of the trackster collection
+  // linkedTrackstersToInputTrackstersMap contains the mapping between the linked tracksters and the input tracksters
+  linkingAlgo_->linkTracksters(input, *resultTracksters, *linkedResultTracksters, linkedTracksterIdToInputTracksterId);
+
+  // Now we need to remove the tracksters that are not linked
+  // We need to emplace_back in the resultTracksters only the tracksters that are linked
+
+  for (auto const &resultTrackster : *resultTracksters) {
+    for (auto const &clusterIndex : resultTrackster.vertices()) {
+      (*resultMask)[clusterIndex] = 0.f;
+    }
+  }
+
+  assignPCAtoTracksters(
+      *resultTracksters, layerClusters, layerClustersTimes, rhtools_.getPositionLayer(rhtools_.lastLayerEE()).z(), true);
+
+  evt.put(std::move(linkedResultTracksters));
+  evt.put(std::move(resultMask));
+  evt.put(std::move(resultTracksters));
+}
+
+void TracksterLinksProducer::printTrackstersDebug(const std::vector<Trackster> &tracksters, const char *label) const {
+#ifdef EDM_ML_DEBUG
+  int counter = 0;
+  for (auto const &t : tracksters) {
+    LogDebug("TracksterLinksProducer")
+        << counter++ << " TracksterLinksProducer (" << label << ") obj barycenter: " << t.barycenter()
```

Comment:
The `#ifdef` could be avoided with
```c++
LogDebug("TracksterLinksProducer").log([&](auto& log) {
  for (auto const &t : tracksters) {
    log  << counter++ << " TracksterLinksProducer (" << label << ") obj barycenter: " << t.barycenter()
    ...
  }
  ...
});
```
In this case you'd also get only one message decoration instead of each `LogDebug()` message having its own decoration.